text/0813-template-layering.md

```diff
@@ -0,0 +1,984 @@
+
+Ember has always had the ability to compile template at runtime. However, since
+this incur significant costs and most apps do not benefit from it, this feature
+is disabled by default and requires and explicit opt-in to include the runtime
+template compiler into the app's bundle. The mechanism of the opt-in will be
+discussed later in this section.
+
+If the template compiler is available at runtime, then it will be possible to
+use the `template()` function at runtime as well. However, as mentioned above,
+this does not extend to the "tagged template string" format. Attempting to use
+the `template()` function as a tag at runtime will result in a helpful error in
+debug builds and undefined behavior on production builds.
+
+If the template compiler is not available at runtime, then attempts to call the
+`template()` function at runtime will result in a helpful error in debug builds
+and undefined behavior on production builds. On production builds, it is not
+guaranteed that the `template` named export may be undefined (i.e. it may not
+even be a function), but developers must not rely on this to "feature-detect"
+the availability of runtime compilation.
+
+Notably, even when template compilation is available at runtime, the result of
+the compilations may be subtly different. This is because that applications may
+have custom glimmer/handlebars AST plugins in their build, and these plugins
+will not be available at runtime.
+
+This may change in the future, but it has always been true for runtime template
+compilation and will likely remain to be the case for the foreseeable future,
+as these plugins are often authored or packaged in ways that are inherently not
+browser-friendly. This usually only used for optional syntax extensions such as
+"inline {{let}}" so in practice it is not an issue. However, some polyfills for
+template features also uses the AST plugin API and those polyfilled features
+will not be available to templates compiled at runtime.
+
+#### Opting In
+
+Today, applications can opt-in to bundling the runtime compiler with something
+along these lines in their `ember-cli-build.js`:
+
+```js
+app.import("node_modules/ember-source/dist/ember-template-compiler.js");
+```
+
+The exact details depend on the version of `ember-source` and is not very well
+documented. It also does not align with the direction we are headed, where we
+encourage using the modules system and imports to express dependencies. This
+configuration is also not friendly to tree-shaking/code-splitting, as it forces
+the template compiler to be included into the main/initial bundle, when it may
+only be needed in one of the infrequently used routes.
+
+This RFC takes the opportunity to improve on this by proposing a new module –
+`@ember/template-compilation/runtime`. When this module is imported, either as
+a "side-effect" import statement, or if its exported values are imported, then
+it serves as an opt-in for enabling runtime template compilation.
```

Comment:
```suggestion
This RFC takes the opportunity to improve on this by proposing a new module –
`@ember/template-compilation/runtime`. We will not compile out the `template()`
function imported from this module. Additionally, importing the module as a
"side-effect" import statement will also serve as an opt-in for including it for
dynamic runtime template compilation.
```

(I tripped over the original phrasing here and am suggesting this after chatting about it with @chancancode in Discord.)